tools/SourceKit/lib/SwiftLang/SwiftDocSupport.cpp

```diff
@@ -516,15 +523,18 @@ static void passInheritsAndConformancesForValueDecl(const ValueDecl *VD,
                Consumer);
 }
 
-static void reportRelated(ASTContext &Ctx,
-                          const Decl *D,
-                          const Decl* SynthesizedTarget,
+static void reportRelated(ASTContext &Ctx, const Decl *D,
+                          TypeOrExtensionDecl SynthesizedTarget,
                           DocInfoConsumer &Consumer) {
   if (!D || isa<ParamDecl>(D))
     return;
   if (const auto *ED = dyn_cast<ExtensionDecl>(D)) {
     if (SynthesizedTarget) {
-      passExtends((const ValueDecl*)SynthesizedTarget, Consumer);
+      auto Extends = SynthesizedTarget.dyn_cast<NominalTypeDecl *>();
+      if (!Extends)
```

Comment:
You have this same "figure out the nominal type declaration to which `SynthesizedTarget` refers" pattern in two places; can you pull it out into a function somewhere?

Reply:
Oh, yeah, bit embarrassing that I missed that. I've made `TypeOrExtensionDecl` a wrapper struct with some convenience methods like `getBaseNominal`.

Reply:
Although, it is in a somewhat clunky place: `PrintOptions.h`. Any thoughts for where something like that should go? `Decl.h` seems wrong to me.

Reply:
I suspect that `TypeOrExtensionDecl` could be more widely used, if we were to search around for pointer unions similar to the one you're replacing. It's fine to put it in `Decl.h`

Reply:
FWIW, I couldn't find any other NominalTypeDecl, ExtensionDecl pointer unions.

Reply:
Ugh, moving it to Decl.h is rather ugly: PrintOptions.h doesn't/can't include that file, so everything has to be indirected...